tendermint/src/abci/response.rs

```diff
@@ -0,0 +1,1027 @@
+            InfoResponse::Info(x) => Self::Info(x),
+            InfoResponse::Query(x) => Self::Query(x),
+        }
+    }
+}
+
+impl TryFrom<Response> for InfoResponse {
+    type Error = &'static str;
+    fn try_from(req: Response) -> Result<Self, Self::Error> {
+        match req {
+            Response::Echo(x) => Ok(Self::Echo(x)),
+            Response::Info(x) => Ok(Self::Info(x)),
+            Response::Query(x) => Ok(Self::Query(x)),
+            _ => Err("wrong request type"),
+        }
+    }
+}
+
+/// The snapshot category of ABCI responses.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum SnapshotResponse {
+    /// Returns a list of local state snapshots.
+    ///
+    /// [ABCI documentation](https://docs.tendermint.com/master/spec/abci/abci.html#listsnapshots)
+    ListSnapshots(ListSnapshots),
+    /// Returns the application's response to a snapshot offer.
+    ///
+    /// See also the [`Snapshot`] data type and the [ABCI state sync documentation][ssd].
+    ///
+    /// [ABCI documentation](https://docs.tendermint.com/master/spec/abci/abci.html#offersnapshot)
+    ///
+    /// [ssd]: https://docs.tendermint.com/master/spec/abci/apps.html#state-sync
+    OfferSnapshot(OfferSnapshot),
+    /// Returns a snapshot chunk from the application.
+    ///
+    /// [ABCI documentation](https://docs.tendermint.com/master/spec/abci/abci.html#loadsnapshotchunk)
+    LoadSnapshotChunk(LoadSnapshotChunk),
+    /// Returns the result of applying a snapshot chunk and associated data.
+    ///
+    /// The application can choose to refetch chunks and/or ban P2P peers as
+    /// appropriate. Tendermint will not do this unless instructed by the
+    /// application.
+    ///
+    /// [ABCI documentation](https://docs.tendermint.com/master/spec/abci/abci.html#applysnapshotchunk)
+    ApplySnapshotChunk(ApplySnapshotChunk),
+}
+
+impl From<SnapshotResponse> for Response {
+    fn from(req: SnapshotResponse) -> Self {
+        match req {
+            SnapshotResponse::ListSnapshots(x) => Self::ListSnapshots(x),
+            SnapshotResponse::OfferSnapshot(x) => Self::OfferSnapshot(x),
+            SnapshotResponse::LoadSnapshotChunk(x) => Self::LoadSnapshotChunk(x),
+            SnapshotResponse::ApplySnapshotChunk(x) => Self::ApplySnapshotChunk(x),
+        }
+    }
+}
+
+impl TryFrom<Response> for SnapshotResponse {
+    type Error = &'static str;
+    fn try_from(req: Response) -> Result<Self, Self::Error> {
+        match req {
+            Response::ListSnapshots(x) => Ok(Self::ListSnapshots(x)),
+            Response::OfferSnapshot(x) => Ok(Self::OfferSnapshot(x)),
+            Response::LoadSnapshotChunk(x) => Ok(Self::LoadSnapshotChunk(x)),
+            Response::ApplySnapshotChunk(x) => Ok(Self::ApplySnapshotChunk(x)),
+            _ => Err("wrong request type"),
+        }
+    }
+}
+
+// =============================================================================
+// Protobuf conversions
```

Comment:
Our current convention is to organize all code relating to a specific type together with that type (e.g. all code relating to the `Exception` struct should be located directly underneath the struct, as opposed to anywhere else in the module).

Reply:
I think this kind of functional grouping is clearer -- the protobuf conversion code is pretty repetitive boilerplate, so it's nice to be able to just look at all of the structure definitions together, rather than alternating between boilerplate -- but I'm happy to change it for consistency if you'd prefer.

Reply:
Yes, our current convention is actually to split out each struct into its own file and co-locate all `impl`s along with that struct.

Reply:
Great, I'll split out the `Request` and `Response` variant types.

Reply:
Split out the types into individual files, each with all the proto impls.